etl/snapshot.py

```diff
@@ -117,55 +119,56 @@ def _cached_dvc_repo(self) -> Any:
         else:
             return get_dvc(use_cache=True)
 
+    def _download_dvc_file(self, md5: str) -> None:
+        """Download file from DVC remote to self.path."""
+        self.path.parent.mkdir(exist_ok=True, parents=True)
+        if self.metadata.is_public:
+            https_url = DVC_CONFIG['remote "public-read"']["url"]
+            download_url = f"{https_url}/{md5[:2]}/{md5[2:]}"
+            files.download(download_url, str(self.path), progress_bar_min_bytes=2**100)
+        else:
+            s3_url = DVC_CONFIG['remote "private"']["url"]
+            download_url = f"{s3_url}/{md5[:2]}/{md5[2:]}"
+            s3_utils.download(download_url, str(self.path))
+
     def pull(self, force=True) -> None:
         """Pull file from S3."""
         if not force and not self.is_dirty():
             return
 
-        from dvc.exceptions import CheckoutError
+        assert len(self.metadata.outs) == 1
+        md5 = self.metadata.outs[0]["md5"]
 
-        # DVC locking is terrible. It'd fail if we didn't use our own file locks.
-        # This is pretty limiting as it allows us to only pull
-        # one snapshot at a time. One option is to pre-pull all snapshot steps or
-        # we could just download the file directly.
-        # The combination of different kinds of locks is kinda magical and waiting
-        # to be refactored soon.
-        with dvc_pull_lock, _unignore_backports(self.path):
-            try:
-                repo = self._cached_dvc_repo()
-                repo.pull(str(self.path), remote="public-read" if self.metadata.is_public else "private", force=force)
-            except CheckoutError as e:
-                raise Exception(
-                    "File not found in DVC. Have you run the snapshot script? Make sure you're not using `is_public: false`."
-                ) from e
+        self._download_dvc_file(md5)
+
+        # update metadata
+        with open(self.metadata_path, "r") as f:
+            m = ruamel_load(f)
+
+        m["outs"][0]["md5"] = checksum_file(self.path)
+        m["outs"][0]["size"] = self.path.stat().st_size
+        m["outs"][0]["path"] = self.path.name
+
+        with open(self.metadata_path, "w") as f:
+            f.write(ruamel_dump(m))
```

Comment:
I hadn't come across `ruamel` before. It might be more readable if the `ruamel` methods are `yaml_load` and `yaml_dump` methods instead, but it's only a soft preference.

Reply:
We already have both `ruamel_dump` and `yaml_dump`. ruamel tries its best to minimize modifications to the yaml file, while `yaml_dump` strips all comments and style. Maybe we could unify them one day.